### PR TITLE
Add com.sublimehq.SublimeText (4)

### PR DESCRIPTION
**Pull request by ryanpcmcquen**
Signed-off-by: Ryan McQuen <rpcm@linux.com>

### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**

Requested: https://forum.sublimetext.com/t/requesting-permission-to-submit-flatpak-for-sublime-text-4/62954
- [ ] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
 
I do not own the domain.
- [ ] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

No patches or file submissions necessary.

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

### CODE REVIEW

File: com.sublimehq.SublimeText.appdata.xml
Comment (hfiguiere) [inline com.sublimehq.SublimeText.appdata.xml:4]: this is not consistent with this: https://github.com/flathub/com.sublimetext.three

File: com.sublimehq.SublimeText.desktop
Comment (hfiguiere) [inline com.sublimehq.SublimeText.desktop:7]:
```suggestion
Exec=subl %F
```

File: com.sublimehq.SublimeText.desktop
Comment (hfiguiere) [inline com.sublimehq.SublimeText.desktop:17]:
```suggestion
Exec=subl --launch-or-new-window
```

File: com.sublimehq.SublimeText.desktop
Comment (hfiguiere) [inline com.sublimehq.SublimeText.desktop:22]:
```suggestion
Exec=subl --command new_file
```

File: com.sublimehq.SublimeText.yaml
Comment (hfiguiere) [inline com.sublimehq.SublimeText.yaml:15]:
```suggestion
  - --socket=fallback-x11
  - --socket=wayland
```

assuming this supports wayland

File: com.sublimehq.SublimeText.yaml
Comment (hfiguiere) [inline com.sublimehq.SublimeText.yaml:17]: that's a broad permission

File: com.sublimehq.SublimeText.yaml
Comment (hfiguiere) [inline com.sublimehq.SublimeText.yaml:19]: this is the wrong python version

File: com.sublimehq.SublimeText.yaml
Comment (hfiguiere) [inline com.sublimehq.SublimeText.yaml:85]:
```suggestion
```

you don't really need a post-install in that case.

File: com.sublimehq.SublimeText.desktop
Comment (hfiguiere) [inline com.sublimehq.SublimeText.desktop:10]: this is the wrong icon name

**Review by hfiguiere — CHANGES_REQUESTED** [review]

File: com.sublimehq.SublimeText.appdata.xml
Comment (ryanpcmcquen) [reply com.sublimehq.SublimeText.appdata.xml:4]: Reasoning is discussed here: https://github.com/flathub/com.sublimetext.three/issues/26

Even though people call it Sublime Text 4, it is really just 'Sublime Text'. There are no plans for a major new version, just continual updates to this with a new pricing model.

File: com.sublimehq.SublimeText.yaml
Comment (ryanpcmcquen) [reply com.sublimehq.SublimeText.yaml:16]: It's what the Sublime Text 3 Flatpak has, I am open to suggestions.

File: com.sublimehq.SublimeText.yaml
Comment (ryanpcmcquen) [reply com.sublimehq.SublimeText.yaml:18]: Well, Sublime Text ships with 3.8, and that's what all the plugins run on (unless they are old, in which case there is also a 3.3 version included).

File: com.sublimehq.SublimeText.desktop
Comment (ryanpcmcquen) [reply com.sublimehq.SublimeText.desktop:10]: Sorry, this was the upstream desktop file and I missed a few things. Thanks for the quick feedback!

File: com.sublimehq.SublimeText.appdata.xml
Comment (hfiguiere) [reply com.sublimehq.SublimeText.appdata.xml:4]: but all the links point to `sublimetext.com`.... include the download and the EULA

File: com.sublimehq.SublimeText.yaml
Comment (hfiguiere) [inline com.sublimehq.SublimeText.yaml:50]: you mean the text editor wil not honour the numeric format of the user's locale and likely a incorrect number formatting?

File: com.sublimehq.SublimeText.desktop
Comment (hfiguiere) [reply com.sublimehq.SublimeText.desktop:10]: since it's ustream, why is it here?

File: com.sublimehq.SublimeText.yaml
Comment (ryanpcmcquen) [reply com.sublimehq.SublimeText.yaml:50]: I'm not sure, this was on Sublime Text 3 so I figured it was necessary, I can test without.

File: com.sublimehq.SublimeText.appdata.xml
Comment (ryanpcmcquen) [reply com.sublimehq.SublimeText.appdata.xml:4]: I could include links from here:
https://www.sublimehq.com/

File: com.sublimehq.SublimeText.desktop
Comment (ryanpcmcquen) [reply com.sublimehq.SublimeText.desktop:10]: I wanted something pure to start with.

**Comment by hfiguiere** [comment]
bot, build com.sublimehq.SublimeText

File: com.sublimehq.SublimeText.desktop
Comment (ryanpcmcquen) [reply com.sublimehq.SublimeText.desktop]: Fixed!

File: com.sublimehq.SublimeText.yaml
Comment (ryanpcmcquen) [reply com.sublimehq.SublimeText.yaml:50]: @ptomato do you know more about this?

File: com.sublimehq.SublimeText.appdata.xml
Comment (ryanpcmcquen) [reply com.sublimehq.SublimeText.appdata.xml:3]: It appears I had the issues link incorrect, it lives under the `sublimehq` GitHub:
https://github.com/sublimehq/

**Comment by ryanpcmcquen** [comment]
Fixed the icon paths.

**Comment by ryanpcmcquen** [comment]
bot, build com.sublimehq.SublimeText

File: com.sublimehq.SublimeText.yaml
Comment (hfiguiere) [reply com.sublimehq.SublimeText.yaml:16]: I would say `xdg-documents` but I'm sure not everyone would agree. Does Sublime Text use portals ? it should.

Comment (bilelmoussaoui) [reply com.sublimehq.SublimeText.yaml:16]: It is a "ide", xdg-documents will make it useless for everyone

Comment (ryanpcmcquen) [reply com.sublimehq.SublimeText.yaml:16]: I use Sublime Text to edit everything, including stuff that requires `sudo`, if we reduce that functionality, I imagine no one will be very happy about it.

**Comment by ryanpcmcquen** [comment]
Looks good here, installing Package Control works. Editing and saving files works. Shows up in 'Applications' for Gnome. How do I test this with `sudo`, new to Flatpak and the path to run this isn't clear.

File: com.sublimehq.SublimeText.yaml
Comment (ptomato) [reply com.sublimehq.SublimeText.yaml:16]: I'm not sure whether it uses portals. Sublime Text 3 doesn't, in any case, and is unlikely to be updated to do so now that Sublime Text 4 exists. Could we test whether it does use portals and then use that information to make a decision?

Comment (ryanpcmcquen) [reply com.sublimehq.SublimeText.yaml:16]: I don't know how to test that.

**Comment by ryanpcmcquen** [comment]
Test for portals (removes `filesystem=host` permission).

**Comment by ryanpcmcquen** [comment]
bot, build com.sublimehq.SublimeText

File: com.sublimehq.SublimeText.yaml
Comment (ryanpcmcquen) [reply com.sublimehq.SublimeText.yaml]: It must use portals, because I removed the permission and I am still able to open, edit, and save files.

Comment (ryanpcmcquen) [reply com.sublimehq.SublimeText.yaml]: The only weird thing is that it shows the wrong path in the title bar now, which I would find unbelievably annoying in daily use.

**Comment by ryanpcmcquen** [comment]
Screenshot from 2022-02-17 20-54-27
Without the `filesystem=host` permission the title bar shows weird path for the file names. The files do get edited and saved in the correct spot, but this would be really annoying for daily use.

File: com.sublimehq.SublimeText.yaml
Comment (ryanpcmcquen) [reply com.sublimehq.SublimeText.yaml]: More details: https://github.com/flathub/flathub/pull/2855#issuecomment-1043888847

**Comment by ryanpcmcquen** [comment]
bot, build com.sublimehq.SublimeText

**Comment by glow12121** [comment]
> ```
> flatpak install --user https://dl.flathub.org/build-repo/76849/com.sublimehq.SublimeText.flatpakref
> ```

Just for everyone's information: for the program to appear on the Start menu, you need to reboot. At least this is so on my Solus Linux setup.

**Comment by ryanpcmcquen** [comment]
> > ```
> > flatpak install --user https://dl.flathub.org/build-repo/76849/com.sublimehq.SublimeText.flatpakref
> > ```
> 
> 
> Just for everyone's information: for the program to appear on the Start menu, you need to reboot. At least this is so on my Solus Linux setup.

Is that Budgie? It shows up right away in the Gnome launcher.

**Comment by glow12121** [comment]
@ryanpcmcquen, for some reason ST does not have a little star on the bottom panel of the Budgie desktop environment. Because of that ST cannot be made a "favorite" program and stay on this panel. As if some permission were missing - that's just my guess though.

Screenshot from 2022-02-21 19-02-46

Screenshot from 2022-02-21 19-03-02

**Comment by glow12121** [comment]
> Is that Budgie? It shows up right away in the Gnome launcher.

Budgie, yes. In my case it didn't show up on the Start menu. Not sure what Gnome Launcher is.

I also tried launching it in the terminal but there was no command starting with "sublime..."

**Comment by ryanpcmcquen** [comment]
The command is: `subl`

But it won't be in your path using the `--user` flag in the install command here. If you omit `--user` from the above command you'll see the expected behavior.

**Comment by glow12121** [comment]
@ryanpcmcquen, thanks. Would you recommend reinstalling ST this way? Could you explain how? Will it likely enable the bottom menu stars, too?

**Comment by ryanpcmcquen** [comment]
> @ryanpcmcquen, thanks. Would you recommend reinstalling ST this way? Could you explain how? Will it likely enable the bottom menu stars, too?

Yes, but I would wait for this to get merged.

**Comment by ryanpcmcquen** [comment]
@hfiguiere got permission from the Sublime HQ team to submit this. Anything left to do?

https://forum.sublimetext.com/t/requesting-permission-to-submit-flatpak-for-sublime-text-4/62954

**Comment by ryanpcmcquen** [comment]
bot, build com.sublimehq.SublimeText

**Comment by barthalion** [comment]
bot, build com.sublimehq.SublimeText

File: com.sublimehq.SublimeText.appdata.xml
Comment (barthalion) [inline com.sublimehq.SublimeText.appdata.xml:14]: This tag is not supported. Please change the entire disclaimer to something like [Discord does](https://flathub.org/apps/details/com.discordapp.Discord). The description itself should be also expanded.

**Comment by ryanpcmcquen** [comment]
bot, build com.sublimehq.SublimeText

**Comment by ryanpcmcquen** [comment]
bot, build com.sublimehq.SublimeText

File: com.sublimehq.SublimeText.yaml
Comment (barthalion) [inline com.sublimehq.SublimeText.yaml:51]: Please add configuration for https://github.com/flathub/flatpak-external-data-checker/ to make sure it gets automatically updated. When it's in, it's good to be merged.

**Comment by barthalion** [comment]
It is a wrapper. We don't redistribute Sublime, just a script that will tell flatpak to download it from elsewhere and process locally, initiated by the user on their computer. The disclaimer and its location has been already approved by Discord, Jetbrains and other vendors, so I don't think Sublime team would have a reason to complain.

**Comment by barthalion** [comment]
It's a hill I'm willing to die on. There's nothing wrong with the form and placement of this disclaimer, and it's somewhat standard across Flathub by now. There are over 50 applications using it, from both big and small developers.

**Comment by ryanpcmcquen** [comment]
> It's a hill I'm willing to die on. There's nothing wrong with the form and placement of this disclaimer, and it's somewhat standard across Flathub by now. There are over 50 applications using it, from both big and small developers.

Can we make the text bold or all caps at least?

**Comment by barthalion** [comment]
This is not currently supported, but we're getting there. There's a PR for flatpak-builder that will switch it to libappstream which supports all the good parts.

Eventually, we will probably display a similar disclaimer under all applications that are not verified by default.

**Comment by ryanpcmcquen** [comment]
> This is not currently supported, but we're getting there. There's a PR for flatpak-builder that will switch it to libappstream which supports all the good parts.
> 
> Eventually, we will probably display a similar disclaimer under all applications that are not verified by default.

I don't feel comfortable submitting it with such a quiet disclaimer. I know it is subjective, but this is not 'clear' to me as an end user. I honestly don't think enough people will read it with the proposed formatting and placement.

If someone wants to take this over who feels differently, that is fine, but I am going to wait for FlatHub to have stronger categorization and/or disclaimers.

**Comment by ptomato** [comment]
It seems conflicting goals have led to an impasse here — @ryanpcmcquen doesn't feel that the standard disclaimer would fulfill the spirit of what Sublime HQ has asked, and doesn't wish to continue otherwise. @barthalion I'm not sure why you are willing to die on this hill but I'm guessing it's because having a nonstandard disclaimer would weaken the existing standard disclaimers? (Correct me if I'm wrong.)

Maybe it would satisfy both goals if someone were to write a small Sublime plugin that displays a markdown document containing the same text as the nonstandard disclaimer, the first time the editor is opened, and pre-install it?

**Comment by ryanpcmcquen** [comment]
> It seems conflicting goals have led to an impasse here — @ryanpcmcquen doesn't feel that the standard disclaimer would fulfill the spirit of what Sublime HQ has asked, and doesn't wish to continue otherwise. @barthalion I'm not sure why you are willing to die on this hill but I'm guessing it's because having a nonstandard disclaimer would weaken the existing standard disclaimers? (Correct me if I'm wrong.)
> 
> Maybe it would satisfy both goals if someone were to write a small Sublime plugin that displays a markdown document containing the same text as the nonstandard disclaimer, the first time the editor is opened, and pre-install it?

I like this idea, and can write such a plugin, if the room feels that is an acceptable compromise.

**Comment by barthalion** [comment]
@ptomato Mostly because we already have a disclaimer that has been proven to satisfy various vendors. The idea of showing such disclaimer at launch is a good idea, and could also be used to show tips&tricks like flatpaked VS Code does.

**Comment by dm17** [comment]
Would be great to have this. Isn't §2b evidence that we're fine here:
`"Only the Linux version of the SOFTWARE PRODUCT may be distributed to third parties"` 
https://www.sublimehq.com/eula

**Comment by dm17** [comment]
@ryanpcmcquen Doesn't this ^ solve the problem?

**Comment by ryanpcmcquen** [comment]
No. I no longer use Flatpak but if someone would be able to write a small plugin that would show a disclaimer on first launch we would be golden.

**Comment by x80486** [comment]
Maintainers stated why the Flatpak version wouldn't work completely end-to-end. I think that may be true, but probably something that could be solved if some changes are implemented on their end. Would be great if someone with (advance) knowledge in this Flatpak business could chime in and confirm if what they describe could be solved in the long run, or not.

Again, none of that is related with the intrinsically difficult problem around displaying something at startup :smirk:

File: com.sublimehq.SublimeText.yaml
Comment (hfiguiere) [inline com.sublimehq.SublimeText.yaml:5]: this runtime will be EOL end of August. You shoud at least use 22.08.

File: com.sublimehq.SublimeText.yaml
Comment (hfiguiere) [inline com.sublimehq.SublimeText.yaml:81]:
```suggestion
      - mkdir -p /app/lib
      - install -Dm755 apply_extra -t /app/bin
      - install -Dm755 subl /app/bin/subl
```

**Comment by ryanpcmcquen** [comment]
I stopped using Flatpak, if someone wants to take this over and display a warning at first launch, go for it.

**Pull request closed without merging**
